imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -211,6 +213,52 @@ def set_spin_bin(l1b_de: xr.Dataset, spin_angle: np.ndarray) -> xr.Dataset:
     return l1b_de
 
 
+def set_spin_cycle(l1a_de: xr.Dataset, l1b_de: xr.Dataset) -> xr.Dataset:
+    """
+    Set the spin cycle for each direct event.
+
+    spin_cycle = spin_start + 7 + (esa_step - 1) * 2
+
+    where spin_start is the spin number for the first spin
+    in an Aggregated Science Cycle (first spin number of an epoch
+    in the L1A spin dataset) and esa_step is the esa_step for a direct event
+
+    Parameters
+    ----------
+    l1a_de : xarray.Dataset
+        The L1A DE dataset.
+    l1b_de : xarray.Dataset
+        The L1B DE dataset.
+
+    Returns
+    -------
+    l1b_de : xarray.Dataset
+        The L1B DE dataset with the spin cycle added for each direct event.
+    """
+    counts = l1a_de["de_count"].values
+    # split the esa_steps into ASC groups
+    de_asc_groups = np.split(l1a_de["esa_step"].values, np.cumsum(counts)[:-1])
+    spin_cycle = []
+    for i, esa_asc_group in enumerate(de_asc_groups):
+        # TODO: Spin Number does not reset for each pointing. Need to figure out
+        #  how to retain this information across days
+        # increment the spin_start by 28 after each aggregated science cycle
+        spin_start = i * 28
+        # calculate the spin cycle for each DE in the ASC group
+        # TODO: Add equation number in algorithm document when new version is
+        # available. Add to docstring as well
+        spin_cycle.extend(spin_start + 7 + (esa_asc_group - 1) * 2)
```

Comment:
This was a bit tricky to follow what was going on, but looks good.

One alternative suggestion instead of the split/cumsum combination would be to create an empty spin_cycle variable initially in your current dataset, and then do something like iterate through the subgroups by downselecting each spin and then updating that subsetted array in place.

```python
for group_start, group_end in ...:
    group_ds = ds.isel(epoch=slice(group_start, group_end))
    group_ds["spin_cycle"] = spin_start + 7 + (group_ds["esa_step"] - 1) * 2)
```

Reply:
I agree this is better. I'm having trouble getting this to work and want to move on to the next PR, so I wrote an issue to make this update in the future: #1556